core/shared/src/main/scala/scalaz/zio/internal/OneShot.scala

```diff
@@ -0,0 +1,43 @@
+// Copyright (C) 2018 John A. De Goes
+package scalaz.zio.internal
+
+/**
+ * A variable that can be set a single time. The synchronous,
+ * effectful equivalent of `Promise`.
+ */
+private[zio] class OneShot[A] private (@volatile var value: A) {
```

Comment:
since `synchronized` is used in the implementation, is this declared non-private and volatile so the value can be fetched without blocking?

Reply:
`synchronized` ensures exclusive access to state modification (it's used like a lock here) and that listeners can receive notification when the state is changed, while `@volatile` ensures all threads access a consistent version of the state, rather than a stale copy.

Reply:
Got it, this makes more sense to me with the optimization in `get` you made